collector/netclass_rtnl_linux.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2022 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !nonetclass && linux
+// +build !nonetclass,linux
+
+package collector
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"regexp"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/jsimonetti/rtnetlink"
+	"github.com/mdlayher/ethtool"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	netclassRTNLIgnoredDevices = kingpin.Flag("collector.netclass_rtnl.ignored-devices", "Regexp of net devices to ignore for netclass_rtnl collector.").Default("^$").String()
+	netclassRTNLWithStats      = kingpin.Flag("collector.netclass_rtnl.with-stats", "Expose the statistics for each network device, replacing netdev collector.").Bool()
+	operstateStr               = []string{
+		"unknown", "notpresent", "down", "lowerlayerdown", "testing",
+		"dormant", "up",
+	}
+)
+
+type netClassRTNLCollector struct {
+	subsystem             string
+	ignoredDevicesPattern *regexp.Regexp
+	metricDescs           map[string]*prometheus.Desc
+	logger                log.Logger
+}
+
+func init() {
+	registerCollector("netclass_rtnl", defaultDisabled, NewNetClassRTNLCollector)
```

Comment:
Rather than make a whole new collector, what if we made this function like the netdev collector. Where we introduced a flag to pick procfs/sysfs vs netlink via a flag (#2509).

This way we can automatically transition users to the new method, while still offering backwards compatibility.